### PR TITLE
always trigger db export when window interval passes

### PR DESCRIPTION
**Pull request by brunocalza**
This PR changes the behavior of when the database changes are exported and uploaded. See: [ENG-784](https://linear.app/tableland/issue/ENG-784/basin-db-streaming-and-confusing-event-creation)

note: by upgrading duckdb, it seems that [ENG-793](https://linear.app/tableland/issue/ENG-793/vaults-cli-stream-commands-parquet-file-format-differs-from-write) got fixed

@dtbuchholz do you mind testing these changes and if you can find any issues?

### CODE REVIEW

File: internal/app/db.go [hunk 1]
```diff
@@ -25,30 +26,218 @@ type Column struct {
...
+		defer func() {
+			if err := db.Close(); err != nil {
+				fmt.Printf("cannot close db %v \n", err)
+			}
+		}()
+		slog.Info("backing up db", "at", exportPath)
+	} else {
+		slog.Info("backing up current db")
+	}
+	var n int
```
Comment (brunocalza) [inline internal/app/db.go:148]: this query helps us figure out if the database is empty or not, so we don't upload an empty parquet file

File: internal/app/test_util.go [hunk 2]
```diff
@@ -25,7 +25,7 @@ type testRow struct {
 }
 
 func importLocalDB(t *testing.T, file *os.File) *sql.Rows {
-	db, err := sql.Open("duckdb", path.Join(t.TempDir(), "temp.db"))
+	db, err := sql.Open("duckdb", path.Join(t.TempDir(), "tmp.db"))
```
Comment (brunocalza) [inline internal/app/test_util.go:28]: for some reason i had to do this for tests to pass

File: internal/app/db.go [hunk 3]
```diff
@@ -5,14 +5,15 @@ import (
 	"database/sql"
```
Comment (brunocalza) [inline internal/app/db.go:5]: most of the changes in this file are just moving functions around, that's why the diff is big. but the main changes are:
- introduces a goroutine that triggers `replace` when window interval ticks
- adds a check to see if database is empty

Comment (avichalp) [reply internal/app/db.go:5]: sounds good!

**Review by avichalp — APPROVED** [review]
looks good!

**Comment by dtbuchholz** [comment]
@brunocalza this worked great / exactly how i'd expect it to work! 

for reference, this is a log for a stream with a 5 second window:
```
> vaults stream \
--private-key "$(cat .wallet)" \
test_pg.stream_5
2024/03/01 10:24:37 INFO created new db at=/Users/dtb/.vaults/test_pg.stream_5/1709317477559109000.db
2024/03/01 10:24:37 INFO applying create query="CREATE TABLE IF NOT EXISTS stream_5 (id integer,val varchar,fl double)"
2024/03/01 10:24:37 INFO Logical replication started slot=basin_stream_5
2024/03/01 10:24:39 INFO new transaction received
2024/03/01 10:24:39 INFO replaying query="insert into stream_5 (id, val, fl) values (1, 'test', 1.1)"
2024/03/01 10:24:39 INFO transaction acked
2024/03/01 10:24:42 INFO window interval passed
2024/03/01 10:24:42 INFO backing up current db
2024/03/01 10:24:42 INFO closing current db
2024/03/01 10:24:45 INFO deleting db parquet export at=/Users/dtb/.vaults/test_pg.stream_5/1709317477559109000.db.parquet
2024/03/01 10:24:45 INFO deleting db dump at=/Users/dtb/.vaults/test_pg.stream_5/1709317477559109000.db
2024/03/01 10:24:45 INFO deleting db wal at=/Users/dtb/.vaults/test_pg.stream_5/1709317477559109000.db.wal
2024/03/01 10:24:45 INFO created new db at=/Users/dtb/.vaults/test_pg.stream_5/1709317485923334000.db
2024/03/01 10:24:45 INFO applying create query="CREATE TABLE IF NOT EXISTS stream_5 (id integer,val varchar,fl double)"
2024/03/01 10:24:47 INFO window interval passed
2024/03/01 10:24:47 INFO backing up current db
2024/03/01 10:24:47 INFO closing current db
2024/03/01 10:24:47 INFO deleting db dump at=/Users/dtb/.vaults/test_pg.stream_5/1709317485923334000.db
2024/03/01 10:24:47 INFO deleting db wal at=/Users/dtb/.vaults/test_pg.stream_5/1709317485923334000.db.wal
2024/03/01 10:24:47 INFO created new db at=/Users/dtb/.vaults/test_pg.stream_5/1709317487620885000.db
2024/03/01 10:24:47 INFO applying create query="CREATE TABLE IF NOT EXISTS stream_5 (id integer,val varchar,fl double)"
2024/03/01 10:24:52 INFO window interval passed
2024/03/01 10:24:52 INFO backing up current db
2024/03/01 10:24:52 INFO closing current db
2024/03/01 10:24:52 INFO deleting db dump at=/Users/dtb/.vaults/test_pg.stream_5/1709317487620885000.db
2024/03/01 10:24:52 INFO deleting db wal at=/Users/dtb/.vaults/test_pg.stream_5/1709317487620885000.db.wal
2024/03/01 10:24:52 INFO created new db at=/Users/dtb/.vaults/test_pg.stream_5/1709317492619785000.db
2024/03/01 10:24:52 INFO applying create query="CREATE TABLE IF NOT EXISTS stream_5 (id integer,val varchar,fl double)"
2024/03/01 10:24:57 INFO window interval passed
2024/03/01 10:24:57 INFO backing up current db
2024/03/01 10:24:57 INFO closing current db
2024/03/01 10:24:57 INFO deleting db dump at=/Users/dtb/.vaults/test_pg.stream_5/1709317492619785000.db
2024/03/01 10:24:57 INFO deleting db wal at=/Users/dtb/.vaults/test_pg.stream_5/1709317492619785000.db.wal
2024/03/01 10:24:57 INFO created new db at=/Users/dtb/.vaults/test_pg.stream_5/1709317497620338000.db
2024/03/01 10:24:57 INFO applying create query="CREATE TABLE IF NOT EXISTS stream_5 (id integer,val varchar,fl double)"
2024/03/01 10:24:59 INFO new transaction received
2024/03/01 10:24:59 INFO replaying query="insert into stream_5 (id, val, fl) values (2, 'test', 1.2)"
2024/03/01 10:24:59 INFO transaction acked
2024/03/01 10:25:02 INFO window interval passed
2024/03/01 10:25:02 INFO backing up current db
2024/03/01 10:25:02 INFO closing current db
2024/03/01 10:25:05 INFO deleting db parquet export at=/Users/dtb/.vaults/test_pg.stream_5/1709317497620338000.db.parquet
2024/03/01 10:25:05 INFO deleting db dump at=/Users/dtb/.vaults/test_pg.stream_5/1709317497620338000.db
2024/03/01 10:25:05 INFO deleting db wal at=/Users/dtb/.vaults/test_pg.stream_5/1709317497620338000.db.wal
2024/03/01 10:25:05 INFO created new db at=/Users/dtb/.vaults/test_pg.stream_5/1709317505768314000.db
2024/03/01 10:25:05 INFO applying create query="CREATE TABLE IF NOT EXISTS stream_5 (id integer,val varchar,fl double)"
2024/03/01 10:25:07 INFO window interval passed
2024/03/01 10:25:07 INFO backing up current db
2024/03/01 10:25:07 INFO closing current db
2024/03/01 10:25:07 INFO deleting db dump at=/Users/dtb/.vaults/test_pg.stream_5/1709317505768314000.db
2024/03/01 10:25:07 INFO deleting db wal at=/Users/dtb/.vaults/test_pg.stream_5/1709317505768314000.db.wal
2024/03/01 10:25:07 INFO created new db at=/Users/dtb/.vaults/test_pg.stream_5/1709317507621172000.db
2024/03/01 10:25:07 INFO applying create query="CREATE TABLE IF NOT EXISTS stream_5 (id integer,val varchar,fl double)"
2024/03/01 10:25:12 INFO window interval passed
2024/03/01 10:25:12 INFO backing up current db
2024/03/01 10:25:12 INFO closing current db
2024/03/01 10:25:12 INFO deleting db dump at=/Users/dtb/.vaults/test_pg.stream_5/1709317507621172000.db
2024/03/01 10:25:12 INFO deleting db wal at=/Users/dtb/.vaults/test_pg.stream_5/1709317507621172000.db.wal
2024/03/01 10:25:12 INFO created new db at=/Users/dtb/.vaults/test_pg.stream_5/1709317512621212000.db
2024/03/01 10:25:12 INFO applying create query="CREATE TABLE IF NOT EXISTS stream_5 (id integer,val varchar,fl double)"
```

**Comment by dtbuchholz** [comment]
oh, and i can confirm the db streaming parquet file format now works with the `parquet` CLI tool i was previously using. thus, basin file uploads and db streams have the same parquet format 👍

**Pull request merged**